src/ir/module-splitting.cpp

```diff
@@ -90,81 +88,118 @@ namespace {
 
 template<class F> void forEachElement(Table& table, F f) {
   for (auto& segment : table.segments) {
-    assert(segment.offset->is<Const>() &&
-           "TODO: handle non-constant segment offsets");
-    Index baseOffset = segment.offset->cast<Const>()->value.geti32();
+    Name base = "";
+    Index offset = 0;
+    if (auto* c = segment.offset->dynCast<Const>()) {
+      offset = c->value.geti32();
+    } else if (auto* g = segment.offset->dynCast<GlobalGet>()) {
+      base = g->name;
+    }
     for (size_t i = 0; i < segment.data.size(); ++i) {
-      f(Index(baseOffset + i), segment.data[i]);
+      f(base, offset + i, segment.data[i]);
     }
   }
 }
 
 struct TableSlotManager {
+  struct Slot {
+    // If `global` is empty, then this slot is at a statically known index.
+    Name global;
+    Index index;
+
+    // Generate code to compute the index of this table slot
+    Expression* makeExpr(Module& module);
+  };
   Module& module;
   Table& table;
   Table::Segment* activeSegment = nullptr;
-  Index activeBase = 0;
-  std::map<Name, Index> funcIndices;
+  Slot activeBase = {"", 0};
```

Comment:
may be nicer to remove the `= ..` stuff and add a ` = 0` in the struct definition? (the Name is initialized anyhow)